src/ngx_http_lua_ssl_session_fetchby.c

```diff
@@ -376,10 +361,110 @@ ngx_http_lua_ssl_sess_fetch_handler(ngx_ssl_conn_t *ssl_conn,
         ngx_http_lua_close_fake_connection(fc);
     }
 
+    return res;
+}
+
+
+#if HAVE_SSL_CLIENT_HELLO_CB_SUPPORT
+int
+ngx_http_lua_ssl_client_hello_handler(ngx_ssl_conn_t *ssl_conn,
+    int *al, void *arg)
+{
+    int                len;
+    ngx_int_t          rc;
+    const u_char      *id;
+
+    len = SSL_client_hello_get0_session_id(ssl_conn, &id);
+
+    if (len <= 0) {
+        return SSL_CLIENT_HELLO_SUCCESS;
+    }
+
+    rc = ngx_http_lua_ssl_sess_fetch_helper(ssl_conn, id, len);
+
+    if (rc == NGX_AGAIN) {
+        return SSL_CLIENT_HELLO_RETRY;
+    }
+
+    return SSL_CLIENT_HELLO_SUCCESS;
+}
+
+
+ngx_ssl_session_t *
+ngx_http_lua_ssl_sess_fetch_handler(ngx_ssl_conn_t *ssl_conn,
+    const u_char *id, int len, int *copy)
+{
+    ngx_connection_t                *c;
+    ngx_http_lua_ssl_ctx_t          *cctx;
+
+    /* set copy to 0 as we expect OpenSSL to handle
+     * the memory of returned session */
+
+    *copy = 0;
+
+    c = ngx_ssl_get_connection(ssl_conn);
+
+    cctx = ngx_http_lua_ssl_get_ctx(c->ssl->connection);
+
+    if (cctx && cctx->done) {
+        return cctx->session;
+    }
+
     return NULL;
 }
 
 
+#else
+
+/* cached session fetching callback to be set with SSL_CTX_sess_set_get_cb */
+ngx_ssl_session_t *
+ngx_http_lua_ssl_sess_fetch_handler(ngx_ssl_conn_t *ssl_conn,
+#if OPENSSL_VERSION_NUMBER >= 0x10100003L
+    const
+#endif
+    u_char *id, int len, int *copy)
+{
+    ngx_int_t                        rc;
+    ngx_connection_t                *c;
+    ngx_http_lua_ssl_ctx_t          *cctx;
+
+    /* set copy to 0 as we expect OpenSSL to handle
+     * the memory of returned session */
+
+    *copy = 0;
+
+    rc = ngx_http_lua_ssl_sess_fetch_helper(ssl_conn, id, len);
+
+    if (rc == NGX_AGAIN) {
+
+#ifdef SSL_ERROR_PENDING_SESSION
+
+        return SSL_magic_pending_session_ptr();
+
+#else
+
```

Comment:
style: do we need an extra line here? Many precedents in the codebase suggest we don't.

Reply:
@thibaultcha 
But `#ifdef SSL_ERROR_PENDING_SESSION` adds an extra blank line below?